src/WebJobs.Extensions.DurableTask/OutOfProcMiddleware.cs

```diff
@@ -740,6 +821,178 @@ private static bool TryExtractSerializedFailureDetailsFromException(string excep
             return false;
         }
 
+        // Reconstruct a FailureDetails instance from the worker's JSON payload,
+        // recursively converting Properties to native .NET types.
+        private static FailureDetails? BuildFailureDetailsFromJson(JObject obj)
+        {
+            string errorType = obj["ErrorType"]?.Value<string>() ?? string.Empty;
+            string errorMessage = obj["ErrorMessage"]?.Value<string>() ?? string.Empty;
+            string? stackTrace = obj["StackTrace"]?.Value<string>();
+            bool isNonRetriable = obj["IsNonRetriable"]?.Value<bool>() ?? false;
+
+            // Build Properties.
+            IDictionary<string, object?>? properties = null;
+            if (obj["Properties"] is JObject props)
+            {
+                properties = new Dictionary<string, object?>();
+                foreach (var kvp in props)
+                {
+                    if (kvp.Value is JObject value)
+                    {
+                        properties[kvp.Key] = ExtractValue(value);
+                    }
+                }
+            }
+
+            // Parse innder failure details recurtively.
+            FailureDetails? inner = null;
+            if (obj["InnerFailure"] is JObject innerObj)
+            {
+                inner = BuildFailureDetailsFromJson(innerObj);
+            }
+
+            return new FailureDetails(errorType, errorMessage, stackTrace, inner, isNonRetriable, properties);
+        }
+
+        public static IDictionary<string, object?>? ExtractPropertiesFromExceptionJson(string json)
+        {
+            var result = new Dictionary<string, object?>();
+
+            try
+            {
+                JObject? rootNode = JObject.Parse(json);
+                if (rootNode is not JObject rootObj)
+                {
+                    return null; // Root is not a JSON object
+                }
+
+                JObject? propertiesNode = rootObj["Properties"] as JObject;
+                if (propertiesNode == null)
+                {
+                    return null; // "Properties" section missing or not an object
+                }
+
+                foreach (var kvp in propertiesNode)
+                {
+                    var value = kvp.Value as JObject;
+                    if (value == null)
+                    {
+                        continue;
+                    }
+
+                    result[kvp.Key] = ExtractValue(value);
+                }
+            }
+            catch (JsonReaderException)
+            {
+                // Invalid JSON format
+                return null;
+            }
+            catch (Exception)
+            {
+                return null;
+            }
+
+            return result;
+        }
+
+        // Convert a JSON representation of Google.Protobuf.WellKnownTypes.Value into a native .NET value.
```

Comment:
Where and why are we converting protobuf types to JSON? This is not something I would have expected.